cli_tools/gce_vm_image_import/importer/source.go

```diff
@@ -0,0 +1,198 @@
+//  Copyright 2020 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package importer
+
+import (
+	"compress/gzip"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/domain"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/param"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/storage"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/daisycommon"
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+)
+
+// A resource that can be imported to GCE disk images. If an instance of this
+// interface exists, it is expected that validation has already occurred,
+// and that the caller can safely use the resource.
+type resource interface {
+	path() string
+}
+
+// initAndValidateSource takes the sourceFile and sourceImage specified by the user
+// and determines which, if any, is importable. It is an error if both sourceFile and
+// sourceImage are specified.
+func initAndValidateSource(sourceFile, sourceImage string,
+	storageClient domain.StorageClientInterface) (resource, error) {
+	sourceFile = strings.TrimSpace(sourceFile)
+	sourceImage = strings.TrimSpace(sourceImage)
+
+	if sourceFile == "" && sourceImage == "" {
+		return nil, daisy.Errf(
+			"either -source_file or -source_image has to be specified")
+	} else if sourceFile != "" && sourceImage != "" {
+		return nil, daisy.Errf(
+			"either -source_file or -source_image has to be specified, but not both %v %v",
+			sourceFile, sourceImage)
+	}
+
+	if sourceFile != "" {
+		return newFileSource(sourceFile, storageClient)
+	}
+
+	return newImageSource(sourceImage)
+}
+
+// Whether the resource is a file in GCS.
+func isFile(s resource) bool {
+	_, ok := s.(fileSource)
```

Comment:
Minor: my personal preference would be adding a "getSourceType() string" and check source.getSourceType() == SOURCE_TYPE_IMAGE. It's better when there are more source types to handle.

Reply:
My first iteration used this approach, but I figured that if we supported another source in the future it'd be easy to refactor.

Is there another source that you envision? (I couldn't think of any)